course5/homework04/homework04.cc

```diff
@@ -0,0 +1,96 @@
+// Example program
+#include <string>
+#include <cstdio>
+#include <cstdlib>
+
+class Endl
+{
+};
+
+
+class Stream
+{    
+private:
+        std::FILE * file;
+public:
+        static Endl endl;
+
+        Stream()
+        {
+                file = stdout;
+        }
+
+        Stream(const char * filename)
+	{
+                file = std::fopen(filename, "w");
+	}
+
+	~Stream()
+        {
+                std::fclose(file);
+	}
+
+	Stream& operator << (std::string value)
+	{
+		std::fprintf(file, "%s", value.c_str());
+                return *this;
+	}
+
+        Stream& operator << (bool value)
+        {
+                std::fprintf(file, "%s", value ? "true" : "false");
+                return *this;
+        }
+
+        Stream& operator << (int value)
+        {
+                std::fprintf(file, "%d", value);
+                return *this;
+        }
+
+        Stream& operator << (float value)
+        {
+                //std::string format = std::string("%.") + std::to_string(precision).c_str() + std::string("f");
+                std::fprintf(file, "%f", value);
+                return *this;
+        }
+
+        Stream& operator << (double value)
+        {
+                std::fprintf(file, "%f", value);
+                return *this;
+        }
+
+        Stream& operator << (Endl)
+        {
+                std::fprintf(file, "\n");
+                std::fflush(file);
+                return *this;
+        }
+
+};
+
+int main(int argc, char** argv)
+{
+
+        Stream Cout;
+
+        // why: double free or corruption (top) ?
+        if (argc > 1)
+                Cout = Stream(argv[1]);
```

Comment:
Here a temporary Stream object is created, which opens the file named argv[1]. The object get copied to Cout object, effectivly coping the FILE pointer to Cout. After the if statement the temporary Stream object gets destroyed, so it closes the file making the pointer invalid, while the Cout object still has pointer to it.